### ISSUE TITLE
False-positive error: JSC_BAD_FUNCTION_DECLARATION ... in ES5 strict mode

### ISSUE THREAD

**Issue by xtracoder** [google/closure-compiler#2118]
I tried using GCC v20161024 in my project and have got this error:

JSC_BAD_FUNCTION_DECLARATION. functions can only be declared at top level or immediately within another function in ES5 strict mode at <file-path>

I'm compiling with 
  compiler.lang-in = 'ECMASCRIPT5'
  compiler.lang-out = 'ECMASCRIPT5'

And this error seems to be false-positive. I've looked through the code and it seems that there is not check regarding current language in/out in `checkFunctionUse()`
https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/StrictModeCheck.java#L133

**Comment by xtracoder** [comment]
I think the problem is caused by this line
https://github.com/google/closure-compiler/commit/9b4169e6a9a8532d7fd3a463231fb64e4dfe3c9a#diff-260ed5cec9127e9e4572e8d5284c2609L447

https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/DefaultPassConfig.java#L488

- 'strict' checks are unconditionally applied to all js files. 

I see there is 'setStrictModeInput()" was added in https://github.com/google/closure-compiler/commit/73409ece7c4cf94ce9c85bf774564edc519d55a9#diff-e1e96443947a8df3565544b63c8a6649R3157 ... may it need to be used in 'DefaultPassConfig' ?

**Comment by ChadKillingsworth** [comment]
Language modes no longer have an impact on strict mode. To disable validating input for strict mode compliance, use `--jscomp_off=es5Strict` (I think that's the correct one).

**Comment by xtracoder** [comment]
I'm using GCC as an embedded 3rd party library into my library - how should I do this from code?

**Comment by ChadKillingsworth** [comment]
As Java?

**Comment by xtracoder** [comment]
Yes, from java.

My code currently looks like following

    CompilerOptions options = new CompilerOptions();
    options.setLanguageIn(CompilerOptions.LanguageMode.ECMASCRIPT5);

    Compiler compiler = new com.google.javascript.jscomp.Compiler();
    compiler.initOptions(options);
    compiler.compile(externs, input, options);

i tried adding `options.setStrictModeInput(false);` but it did not make any effect :(

**Comment by ChadKillingsworth** [comment]
`options.setWarningLevel(...)`

https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/CompilerOptions.java#L1394-L1399

**Comment by xtracoder** [comment]
So ... I tried

    options.setWarningLevel(DiagnosticGroups.ES5_STRICT, CheckLevel.OFF);

and have got no errors in `compiler.getErrors()` (what is good), but also have got following in build log

    my-js-file.js:140: ERROR - functions can ... in ES5 strict mode
            function a () {
            ^

    1 error(s), 0 warning(s)

It looks completely misleading - build has finished successfully, but there were errors during build ...

**Comment by brad4d** [comment]
@ChadKillingsworth, it's not quite true that language mode no longer affects strictness.
Here are some points that will hopefully clarify things.

-   All of the old-style language modes (ECMASCRIPT5, ECMASCRIPT5_STRICT, etc.) still indicate strictness or non-strictness as they always have.
-   The new-style language modes (ECMASCRIPT_2015, etc.) do not, but the default behavior is to assume strict on input and output, which can be overridden by `--nostrict_mode_input` and `--noemit_use_strict`, respectively.
- The pass for doing strict mode checks is always run. The language mode or `--nostrict_mode_input` option just disable the warnings/errors. Here's the line that does that.
   https://github.com/google/closure-compiler/blob/1fd5b5ca4be6ede6f4c2f1850699f9840a7422d9/src/com/google/javascript/jscomp/Compiler.java#L412

**Comment by brad4d** [comment]
@xtracoder I suspect something is hinky with the way you're running the compiler from Java code.
Perhaps you could share a code snippet showing what you're doing?

**Comment by xtracoder** [comment]
Hinky? Yes, somewhat :)

I've got where the problem is. JS is compiled in 2 stages:

1. extract JsDocs
2. generate minified version

I've applied `setWarningLevel()` only to second phase, it was missing during 1st phase, but since errors are ignored there (no effect on documentation) - it happened that errors were logged from 1st phase and overall success was coming from phase 2. OK, than problem is resolved.

PS: Compilation is split into 2 phase because if I mix

    CompilationLevel.SIMPLE_OPTIMIZATIONS.setOptionsForCompilationLevel(options);

with 

    options.recordFunctionInformation = true;
    options.setParseJsDocDocumentation(INCLUDE_DESCRIPTIONS_WITH_WHITESPACE);

resulting AST already contains **obfuscated** function arguments :( - like 'function myFn(a, b, s)' instead of 'function myFn(param_name1, param_name2, param_name3)' and it is not possible to generate documentation and minify file in one pass.
